public/scripts/power-user.js

```diff
@@ -1825,64 +1841,111 @@ async function loadContextSettings() {
 }
 
 /**
- * Fuzzy search characters by a search term
+ * Generates a hash of the Fuse configuration keys
+ * @param {Array<{name: string, weight: number, getFn?: Function}>} keys
+ * @returns {number} Hash of the keys configuration
+ */
+function hashFuseKeys(keys) {
+    // Convert keys to a stable string representation
+    const keyString = keys.map(k => {
+        const getFnString = k.getFn ? k.getFn.toString() : '';
+        return `${k.name}:${k.weight}:${getFnString}`;
+    }).join('|');
+
+    return getStringHash(keyString);
+}
+
+/**
+ * Common function to perform fuzzy search with caching
+ * @param {string} type - Type of search from fuzzySearchCategories
+ * @param {any[]} data - Data array to search in
+ * @param {Array<{name: string, weight: number, getFn?: Function}>} keys - Fuse.js keys configuration
  * @param {string} searchValue - The search term
  * @returns {import('fuse.js').FuseResult<any>[]} Results as items with their score
  */
-export function fuzzySearchCharacters(searchValue) {
-    // @ts-ignore
-    const fuse = new Fuse(characters, {
-        keys: [
-            { name: 'data.name', weight: 20 },
-            { name: '#tags', weight: 10, getFn: (character) => getTagsList(character.avatar).map(x => x.name).join('||') },
-            { name: 'data.description', weight: 3 },
-            { name: 'data.mes_example', weight: 3 },
-            { name: 'data.scenario', weight: 2 },
-            { name: 'data.personality', weight: 2 },
-            { name: 'data.first_mes', weight: 2 },
-            { name: 'data.creator_notes', weight: 2 },
-            { name: 'data.creator', weight: 1 },
-            { name: 'data.tags', weight: 1 },
-            { name: 'data.alternate_greetings', weight: 1 },
-        ],
+function performFuzzySearch(type, data, keys, searchValue) {
+    const currentKeyHash = hashFuseKeys(keys);
+    const cache = fuzzySearchCaches[type];
+
+    // Check cache for existing results
+    if (cache.keyHash === currentKeyHash && cache.resultMap.has(searchValue)) {
+        // console.debug(`Using cached ${type} fuzzy search results for ${searchValue}`);
+        return cache.resultMap.get(searchValue);
+    }
+
+    // Clear cache if keys changed
+    if (cache.keyHash !== currentKeyHash) {
+        // console.debug(`${type} Fuse keys changed, clearing cache`);
+        cache.keyHash = currentKeyHash;
+        cache.resultMap.clear();
+    }
+
+    const fuse = new Fuse(data, {
+        keys: keys,
         includeScore: true,
         ignoreLocation: true,
         useExtendedSearch: true,
         threshold: 0.2,
     });
 
     const results = fuse.search(searchValue);
-    console.debug('Characters fuzzy search results for ' + searchValue, results);
+    cache.resultMap.set(searchValue, results);
+
+    // console.debug(`${type} fuzzy search results for ${searchValue}`, results);
     return results;
 }
 
+/**
+ * Clears all fuzzy search caches
+ */
+export function clearFuzzySearchCaches() {
```

Comment:
Thank you Cohee! I have no idea what I am doing when it comes to js haha. Let me take a look at Weak collections.

Regarding memory leaks, as long as the application is running search is accessible correct? Say we load a character or a group, shouldn't the cache get invalidated if something changes and otherwise shouldn't the cache remain to increase search performance on the next search?

I'm just trying to understand the use case in which storing this cache wouldn't be beneficial.

Reply:
Looked into Weak collections and they seem like they might not properly work. Would you mind pointing me towards where I can find the loads from backend so I can properly reset the cache? Otherwise I can introduce an LRU cache, maybe that might limit memory and hopefully keep the perf improvement.

Reply:
I placed the call to clear cache on server load but the cache wasn't being cleared often unless we performed an operation that would make the request. 

I've instead placed it at the end of `PrintCharacters` which is called in many places due to it being wrapped by `PrintCharactersDebounced`.

![after_fix_with_cache_clearing](https://github.com/user-attachments/assets/e69c97f1-523f-4470-adfe-d55b3f29f79c)
![after_fix_with_cache_clearing_logs](https://github.com/user-attachments/assets/73e4f7b7-9b90-47ba-99e6-a5e8520fc398)

The performance improvement is still there, increasing now 200ms-360ms but it's still manageable.

I've also removed the key hashing since we're explicitly clearing the cache after printing the characters which will reinitialize it on every search @Wolfsblvt so that should be a more general catch-all to the concern you raised.

Let me know what you guys think.

Reply:
As a side note, I went back and realized I had a bug with hashing the keys as the cache was not being properly updated. I needed to hash the values not the keys but that introduced additional latency.

The search time went from ~350ms to ~800ms with hashing the values. Maybe a bug on my end but it seems that invalidating the cache on `printCharacters` is likely the most performance, and less error prone (given that it gets reset at search time). 

For fun I checked what performance we get if we had some other mechanism to know when characters are updated (i.e calling this on character data save if tags are updated or description is updated) and it's ~150-200ms which looks really fast haha. The only issue is that if we distribute the cache clearing to other areas, it sort of turns into a leaky abstraction given that this is mainly a search cache and now other parts of the system need to know how to use it. 

The most ideal would be having a `lastUpdated` area in any of our data sources (tags, groups, worldInfo, characters) which then we can invalidate the cache by simply checking if cached lastUpdated < current lastUpdated.

Reply:
Disregard, added a comment on new flow.